# Домашнє завдання до теми "Основи роботи з даними у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [1]:
import pandas as pd 
df = pd.read_csv('/Users/kate/Desktop/data/supermarket_sales.csv')
print(type(df))

<class 'pandas.core.frame.DataFrame'>


1. За допомогою бібліотеки Pandas виведіть перші 4 і останні 4 записи у наборі даних (це дві окремі команди). Треба саме 4! Не 5.

In [3]:
print(df.head(4))
print(df.tail(4))

    Invoice ID Branch       City Customer type  Gender  \
0  750-67-8428      A     Yangon        Member  Female   
1  226-31-3081      C  Naypyitaw        Normal  Female   
2  631-41-3108      A     Yangon        Normal    Male   
3  123-19-1176      A     Yangon        Member    Male   

             Product line  Unit price  Quantity   Tax 5%     Total       Date  \
0       Health and beauty       74.69         7  26.1415  548.9715   1/5/2019   
1  Electronic accessories       15.28         5   3.8200   80.2200   3/8/2019   
2      Home and lifestyle       46.33         7  16.2155  340.5255   3/3/2019   
3       Health and beauty       58.22         8  23.2880  489.0480  1/27/2019   

    Time      Payment    cogs  gross margin percentage  gross income  Rating  
0  13:08      Ewallet  522.83                 4.761905       26.1415     9.1  
1  10:29         Cash   76.40                 4.761905        3.8200     9.6  
2  13:23  Credit card  324.31                 4.761905       16.21

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1F5XlzBOOTMsedpsYqz6NnXjIPTDe6jpx)

2. Виведіть рядки даних з індексами з 615 по 619 включно.

In [4]:
print(df[615:620])

      Invoice ID Branch       City Customer type  Gender  \
615  509-29-3912      A     Yangon        Member  Female   
616  327-40-9673      B   Mandalay        Member    Male   
617  840-19-2096      C  Naypyitaw        Member    Male   
618  828-46-6863      A     Yangon        Member    Male   
619  641-96-3695      C  Naypyitaw        Member  Female   

               Product line  Unit price  Quantity   Tax 5%     Total  \
615       Sports and travel       38.72         9  17.4240  365.9040   
616       Sports and travel       72.60         6  21.7800  457.3800   
617  Electronic accessories       87.91         5  21.9775  461.5275   
618      Food and beverages       98.53         6  29.5590  620.7390   
619     Fashion accessories       43.46         6  13.0380  273.7980   

          Date   Time      Payment    cogs  gross margin percentage  \
615  3/20/2019  12:24      Ewallet  348.48                 4.761905   
616  1/13/2019  19:51         Cash  435.60                 4.761

Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1Otufe_TXRmjZxWigzBiVBLCAVUfEReXM)

3. Виведіть назви колонок у наборі даних разом з їхніми типами. Це можна зробити кількома способами, важливо аби в результаті були назви колонок і відповідні їм типи.

In [6]:
print(df.dtypes)

Invoice ID                  object
Branch                      object
City                        object
Customer type               object
Gender                      object
Product line                object
Unit price                 float64
Quantity                     int64
Tax 5%                     float64
Total                      float64
Date                        object
Time                        object
Payment                     object
cogs                       float64
gross margin percentage    float64
gross income               float64
Rating                     float64
dtype: object


Очікуваний результат (один з можливих форматів):

![](https://drive.google.com/uc?export=view&id=1ASh6-S-_f3X9fTQAPIvnHMFZamTBjNkL)

4. Користуючись методом `DataFrame.describe()` виявіть 1, 25, 50, 75 і 99 персентилі колонки `gross income`. Знайдіть, яке значення відповідає 75отму персентилю цієї колонки і що це означає? Як би Ви його пояснили простими словами колегам? Відповідь напишіть текстом в окремій клітинці типу Markdown.

In [7]:
with pd.option_context('display.precision', 2):
    display(df['gross income'].describe(percentiles=[0.01, 0.25, 0.5, 0.75, 0.99]))

count    1000.00
mean       15.38
std        11.71
min         0.51
1%          0.91
25%         5.92
50%        12.09
75%        22.45
99%        45.25
max        49.65
Name: gross income, dtype: float64

Очікуваний результат:

75-й персентиль = 22.45. Тобто 75 % наших доходів  складають 22.45 або менше. Тільки 25% доходів перевищують це значення 

![](https://drive.google.com/uc?export=view&id=1bDa1YFUWDsPm78Lnpy2geQND8M1HEQiT)

5. Виведіть кількість інвойсів за кожним унікальним містом в наборі даних.

  Підказка: інвойс - це один рядок в цьому наборі даних, ми можемо цим скористатись аби швидко отримати відповідь.

In [9]:
print(df.City.value_counts())

City
Yangon       340
Mandalay     332
Naypyitaw    328
Name: count, dtype: int64


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1G8c6M875NOW7oU3OtZegTOnyD2y8lhAV)

6. Виведіть, який відсоток інвойсів припадає на кожне унікальне місто.

In [10]:
print(df.City.value_counts(normalize=True))

City
Yangon       0.340
Mandalay     0.332
Naypyitaw    0.328
Name: proportion, dtype: float64


Очікуваний результат:

![](https://drive.google.com/uc?export=view&id=1yZZpWhL2y8DX9rSRioYCRh1VbJgcf3fY)

7.1. 
Використовуючи фільтрацію даних, знайдіть середнє значення `gross income` для міст `Yangon` та `Mandalay`. В якому місті воно більше?
7.2 Сформулюйте 3 гіпотези, чому в одному місті gross income вище, як можна дізнатись чи правдива ця гіпотеза з використанням даних (наявних і додаткових), та як висновки з цих гіпотез можна використати для прийняття рішень як підвищити прибуток в мережі супермаркетів. Відповідь напишіть текстом в окремій клітинці типу Markdown.

Це завдання - на продуктове мислення. Якщо не знаєте, які гіпотези сформулювати - напишіть ті, які здаються вам зовсім примітивними. Важливо навчитись висловлювати свої гіпотези щодо даних, і з часом та досвідом вони будуть ставати щоразу глибше і точніше. Весь аналіз даних будується на тому, що ми перевіряємо певні гіпотези та робимо висновки і рекоемндації на основі своїх досліджень.

In [13]:
yangon_data = df.query("City == 'Yangon'")['gross income']
mean_gross_income_yangon = yangon_data.sum() / yangon_data.count()
print(mean_gross_income_yangon)

mandalay_data = df.query("City == 'Mandalay'")['gross income']
mean_gross_income_mandalay = mandalay_data.sum() / mandalay_data.count()
print(mean_gross_income_mandalay)


14.874001470588235
15.232024096385542


Середнє значеняя gross income більше у місті Mandalay 
Гіпотези:
- більша платеспроможність начелення і від цього більший чек ( можемо зібрати данні про середній дохід населння ). Якщо данні підтвердяться для підвищення прибутковості мережі додавати преміум сексенти продуктів які могли би зацікавити заможніх покупців 

- в одному місті продається більша кількість товарів чим в іншому (можемо зібрати данні про кількість проданних товарів(аналіз колонки Quantity)). Якщо данні підтвердяться для підвищення прибутковості мережі звернути увагу на маркетинг (проводити рекламні компанії та акційні пропозиції)

- в одному місті є більш популярною якась певна продуктова ліня ( можемо проаналізувати розподіл продажів за продуктовими лініями у кожному місті ) Якщо данні підтвердяться можна розширити асортимент цієї продуктової лінії 

Очікуваний результат:
![](https://drive.google.com/uc?export=view&id=1JdXUmK46WcoXUJYP7NS2SJlr2XjBbpQ8)

8. Для продовження дослідження різниці у прибутках в магазинах давайте зʼясуємо, чи відрізняється маржа ('gross margin percentage') для різних інвойсів в магазині? Перевірте це програмно на даних виходячи з того функціоналу, який ми вже пройшли у лекціях.

Дайте на основі отриманого результату відповідь: "У всіх всіх інвойсах однакова маржа" або "Маржа відрізняється для різних інвойсів".

In [48]:

df_yangon = df[df['City'] == 'Yangon']
df_yangon[['Invoice ID', 'gross margin percentage']]
print('Статистичний опис маржі для Yangon:')
print(df_yangon['gross margin percentage'].describe())


df_mandalay = df[df['City'] == 'Mandalay']
df_mandalay[['Invoice ID', 'gross margin percentage']]
print('\nСтатистичний опис маржі для Mandalay:')
print(df_mandalay['gross margin percentage'].describe())

Статистичний опис маржі для Yangon:
count    3.400000e+02
mean     4.761905e+00
std      8.120828e-15
min      4.761905e+00
25%      4.761905e+00
50%      4.761905e+00
75%      4.761905e+00
max      4.761905e+00
Name: gross margin percentage, dtype: float64

Статистичний опис маржі для Mandalay:
count    3.320000e+02
mean     4.761905e+00
std      7.633690e-15
min      4.761905e+00
25%      4.761905e+00
50%      4.761905e+00
75%      4.761905e+00
max      4.761905e+00
Name: gross margin percentage, dtype: float64


Очікуваний результат: відповідь щодо маржі в усіх інвойсах на основі отриманих результатів

Можемо бачити що дані за статичним описом маржі у місті Yangon є трохи вищими. 
Це може вказувати на різні умови чи стратегії продажу в різних містах

9. Врахуовуючи висновок в попередньому завданні, дослідимо як продаються в цих містах різні категорії товарів. Виведіть відсотковий розпоіділ кількості інвойсів по категоріям товарів `Product line` в містах `Yangon` та `Mandalay`.

In [56]:
yangon_counts = df_yangon['Product line'].value_counts()
yangon_percent = (yangon_counts / yangon_counts.sum())
print(yangon_percent)

mandalay_counts = df_mandalay['Product line'].value_counts()
mandalay_percent = (mandalay_counts / mandalay_counts.sum())
print(mandalay_percent)


Product line
Home and lifestyle        0.191176
Electronic accessories    0.176471
Sports and travel         0.173529
Food and beverages        0.170588
Fashion accessories       0.150000
Health and beauty         0.138235
Name: count, dtype: float64
Product line
Fashion accessories       0.186747
Sports and travel         0.186747
Electronic accessories    0.165663
Health and beauty         0.159639
Food and beverages        0.150602
Home and lifestyle        0.150602
Name: count, dtype: float64


Очікуваний результат для кожного з міст:

![](https://drive.google.com/uc?export=view&id=1IbG58YK5KVqC-wCnllF1btgEgfWTbi_O)

![](https://drive.google.com/uc?export=view&id=1aKOkLmlMq1sOh4GbwyaQyjfhNJzdU4Yh)